File: 346-moving-average-from-data-stream/moving-average-from-data-stream.ts
class MovingAverage {
    private size: number
    private queue: number[]
    private sum: number

    constructor(size: number) {
        this.size = size
        this.queue = []
        this.sum = 0
    }

    next(val: number): number {
        this.queue.push(val)
        this.sum += val

        if (this.queue.length > this.size) {
            const removed = this.queue.shift()
            this.sum -= removed
        }

        return this.sum / this.queue.length
    }
}

/**
 * Your MovingAverage object will be instantiated and called as such:
 * var obj = new MovingAverage(size)
 * var param_1 = obj.next(val)
 */

File: 346-moving-average-from-data-stream/moving-average-from-data-stream.py
class MovingAverage:

    def __init__(self, size: int):
        self.size = size
        self.window_sum = 0
        self.queue = deque()

    def next(self, val: int) -> float:
        self.window_sum += val
        self.queue.append(val)
        if len(self.queue) > self.size:
            self.window_sum -= self.queue.popleft()
        return self.window_sum / len(self.queue)


# Your MovingAverage object will be instantiated and called as such:
# obj = MovingAverage(size)
# param_1 = obj.next(val)